_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+---
+layout: post
+title: A story unfolds
+category: Haskell
+description: or how I tried to plant a tree
+---
+# A story unfolds
+
+### or how I tried to plant a tree
+
+I tried to re-implement the Unix `tree` tool in Haskell, as a learning exercise. The goal is to create a tool that, when provided with a directory, lists all the contents and prints them in a way similar to this:
+```
+$ tree
+.
+├── a
+│   └── c
+└── b
+    ├── d
+    └── e
+
+5 directories, 0 files
+```
+It is well-defined, well-structured, and obviously recursive problem, so it should be easy to do, a no-brainer almost[^excuses].
+
+I started hacking away at the tree.
+
+The thing I would need first was a function to list directory contents, and another one to let me distinguish between directories and files. Quite fortunately, I had a pair of trusty axes on hand[^axes]:
+```haskell
+listDir :: File -> IO [File]
+isDir :: File -> IO Boolean
+```
+A naïve implementation came to mind right away[^madeup]:
+```haskell    
+tree :: File -> IO ()
+tree file = tree' "" file
+  where tree' prefix file = do
+          putStrLn (prefix ++ "- " ++ name file)
+          isDirectory <- isDir file
+          if isDirectory
+          then do
+            contents <- listDir file
+            traverse_ (tree' (' ':prefix)) contents
+          else return ()
+```
+It does not produce nice branch-looking lines between filenames, relying on indentation instead - but it is close enough.
```

Comment:
Heh jus touched a similar thing quite recently ;p https://github.com/VirtusLab/git-machete/commit/7cab1b71fcabd08ac063f3a54f7de9c53b51b9ec#diff-45dd71f96443d6485689b558e3e3db963e6cd98008624b4d7c6e13c56ce486f8R988

Reply:
I did implement the `printTree` using the elbows (including the three-legged ones), but it didn't fit in six lines[^*], so I decided to skip it and write something shorter for the purpose of this post.

[^*]:
    It needed eleven of them.
    Also it turns out you can include footnotes in GitHub comments.

Reply:
> Also it turns out you can include footnotes in GitHub comments

😯 
Actually... how do you do that? I mean, `## Footnotes` is obvious... but how is link for backreference generated?